Templates/Templates/ModelMappable.stencil

```diff
@@ -0,0 +1,162 @@
+    {{ var.name }} = try container.decodeIfPresent({{ typename }}.self, forKey: .{{ var.name }}) ?? ""
+  {% elif typename == "Bool" %}
+    {{ var.name }} = try container.decodeIfPresent({{ typename }}.self, forKey: .{{ var.name }}) ?? false
+  {% elif typename == "Int" or typename == "Double" or typename == "Float" or typename == "NSNumber" %}
+    {{ var.name }} = try container.decodeIfPresent({{ typename }}.self, forKey: .{{ var.name }}) ?? 0
+  {% elif typename == "Date" %}
+    {{ var.name }} = try container.decodeIfPresent({{ typename }}.self, forKey: .{{ var.name }}) ?? Date()
+  {% endif %}
+{% endmacro %}
+
+{% macro decodableInitializer variables %}
+  init(from decoder: Decoder) throws {
+    let container = try decoder.container(keyedBy: CodingKeys.self)
+    {% for var in variables where var|annotated:"incomingType" %}
+    {% call decodeVar var %}
+    {% endfor %}
+    {% for var in variables where var|!annotated:"incomingType" %}
+    {% call decodeWithDefault var %}
+    {% endfor %}
+  }
+{% endmacro %}
+
+{% macro convertVarForObject var %}
+  {% set typename %}{{ var.typeName.name }}{% endset %}
+  {% set itype %}{{ var.annotations.incomingType }}{% endset %}
+  {% if typename == "Bool" and itype == "Int" %} {# its a 0 or 1 #}
+    {{ var.name }} = (object["{{ var.name }}"] as? Int) == 1
+  {% elif typename == "Date" and itype == "String" %} {# iso date string #}
+    if let {{ var.name }} = object["{{ var.name }}"] as? NSNumber {
+      self.{{ var.name }} = Date.fromTimestamp(Timestamp({{ var.name }}.int64Value))
+    } else {
+      self.{{ var.name }} = Date()
+    }
+  {% endif %}
+{% endmacro %}
+
+{% macro convertVarForEncoder var %}
+  {% set typename %}{{ var.typeName.name }}{% endset %}
+  {% set itype %}{{ var.annotations.incomingType }}{% endset %}
+  {% if not itype %}
+    try container.encode({{ var.name }}, forKey: .{{ var.name }})
+  {% elif typename == "Bool" and itype == "Int" %} {# its a 0 or 1 #}
+    try container.encode({{ var.name }} ? 1 : 0, forKey: .{{ var.name }})
+  {% elif typename == "Date" and itype == "String" %} {# iso date string #}
+    try container.encode({{ var.name }}.iso8601, forKey: .{{ var.name }})
+  {% endif %}
+{% endmacro %}
+
+{% macro encode variables %}
+  func encode(to encoder: Encoder) throws {
+    var container = encoder.container(keyedBy: CodingKeys.self)
+    {% for var in variables %}
+    {% call convertVarForEncoder var %}
+    {% endfor %}
+  }
+{% endmacro %}
+
+{% macro asDictionary variables %}
+  func asDictionary() -> [String: Any] {
+    var dictionary: [String: Any] = [:]
+    {% for var in variables %}
+    {% if var.typeName.name == "Date" %}
+    dictionary["{{ var.name }}"] = {{ var.name }}.asTimestamp
+    {% else %}
+    dictionary["{{ var.name }}"] = {{ var.name }}
+    {% endif %}
+    {% endfor %}
+    return dictionary
+  }
+{% endmacro %}
+
+{% for type in types.based.ModelMappable where type|protocol and not type.name == "ModelMappable" %}
```

Comment:
why `ModelMappable`? Other templates use another pattern that will make this `AutoCodable`. I would also split it into `AutoDecodable` and `AutoDecodable`, not always you need encoding, most often you need just decoding.